lib/ansible/config/manager.py

```diff
@@ -65,133 +65,154 @@ def _get_config_label(plugin_type: str, plugin_name: str, config: str) -> str:
                     makedirs_safe(value, 0o700)
+
                 prefix = 'ansible-local-%s' % os.getpid()
                 value = tempfile.mkdtemp(prefix=prefix, dir=value)
                 atexit.register(cleanup_tmp_file, value, warn=True)
-            else:
-                errmsg = 'temppath'
 
-        elif value_type == 'pathspec':
-            if isinstance(value, string_types):
+                return value
+
+        case 'pathspec':
+            if isinstance(value, str):
                 value = value.split(os.pathsep)
 
-            if isinstance(value, Sequence):
-                value = [resolve_path(x, basedir=basedir) for x in value]
-            else:
-                errmsg = 'pathspec'
+            if isinstance(value, Sequence) and not isinstance(value, bytes) and all(isinstance(x, str) for x in value):
+                return [resolve_path(x, basedir=basedir) for x in value]
 
-        elif value_type == 'pathlist':
-            if isinstance(value, string_types):
+        case 'pathlist':
+            if isinstance(value, str):
                 value = [x.strip() for x in value.split(',')]
 
-            if isinstance(value, Sequence):
-                value = [resolve_path(x, basedir=basedir) for x in value]
-            else:
-                errmsg = 'pathlist'
+            if isinstance(value, Sequence) and not isinstance(value, bytes) and all(isinstance(x, str) for x in value):
+                return [resolve_path(x, basedir=basedir) for x in value]
 
-        elif value_type in ('dict', 'dictionary'):
-            if not isinstance(value, Mapping):
-                errmsg = 'dictionary'
+        case 'dictionary' | 'dict':
+            if isinstance(value, dict):
+                return value
 
-        elif value_type in ('str', 'string'):
-            if isinstance(value, (string_types, bool, int, float, complex)):
-                value = to_text(value, errors='surrogate_or_strict')
-                if origin_ftype and origin_ftype == 'ini':
-                    value = unquote(value)
-            else:
-                errmsg = 'string'
+            if isinstance(value, Mapping):
+                return dict(value)
 
-        # defaults to string type
-        elif isinstance(value, (string_types)):
-            value = to_text(value, errors='surrogate_or_strict')
-            if origin_ftype and origin_ftype == 'ini':
-                value = unquote(value)
+        case 'string' | 'str':
+            if isinstance(value, str):
+                return value
 
-        if errmsg:
-            raise ValueError(f'Invalid type provided for {errmsg!r}: {value!r}')
+            if isinstance(value, (bool, int, float, complex)):
+                return str(value)
 
-    return to_text(value, errors='surrogate_or_strict', nonstring='passthru')
+        case _:
+            # FIXME: define and document a pass-through value_type (None, 'raw', 'object', '', ...) and then deprecate acceptance of unknown types
```

Comment:
The 'default' type is 'string', so unspecified should be forced to string, though that was not exactly what was happening (passthrough?). 

So while this departs from what we documented, it is probably 'more correct' reflection of what was happening.